### DISCUSSION TITLE
Usage of Prisma SDK in Rust

### DISCUSSION THREAD

**Discussion by hyochan** [prisma/prisma#11039] [Q&A] [answered]
Today, I am thinking of creating things in Rust with [Prisma SDK](https://www.npmjs.com/package/@prisma/sdk) which is already in npm.

I am wondering if I can have that in Rust without rewriting. Since Prisma Engine is implemented in Rust, I am trying to find out more before I dive into implementing things.
Currently, I find [no crates from Prisma](https://crates.io/search?q=prisma).

What I need from the `sdk` is the `engine-commands` part and I want to find out where to head to.
![image](https://user-images.githubusercontent.com/27461460/148725652-887daa1b-4854-4fac-99cd-057d67b5308d.png)

Thanks in advance.

By the way, Happy new year 🎉

**Comment by tomhoule** [comment, marked as answer]
Hi @hyochan ! There is no official support and no public API for using the engines directly. We make breaking changes to these APIs regularly and without communicating. With all these caveats, if you still want to use the engines from Rust code, I'd suggest using `git` dependencies on https://github.com/prisma/prisma-engines in your Cargo.toml.

I am curious about your use case, how are you trying to use Prisma with Rust?

**Reply by hyochan** [reply]
@tomhoule Hello! Nice to see you again~! Happy new year 🎉 

I am trying to blow up my motivation in using Rust and I was trying to look into where to start from.
Today, I've got some idea on doing that with Prisma. I want to expose common utilities in Rust and support Prisma Generator.

Our team is already doing something with Prisma Schema Generator which is [graphql schema generator](https://github.com/prisma-korea/graphql-schema-generator). Here, we are generally doing some parsing and transpiling with the Prisma DMMF.

I think I can handle that in Rust which would be much more performant as our rule grows.
To add more understanding to what I mean by the `rule` here, you may refer to the https://github.com/prisma-korea/graphql-schema-generator/pull/46#issuecomment-1008221105.

How do you think about the idea and what do you think in general?

**Reply by tomhoule** [reply]
I think it's probably complex but definitely possible. The problem with binding to our internal APIs is that they change pretty often, so you will see breakage, but you can import them like any other regular crate. You can explore the public API of the prisma engines here: https://prisma.github.io/prisma-engines/ — `datamodel` would be the relevant crate for everything schema-related, and `query-engine` for the DMMF.

**Reply by hyochan** [reply]
Thank you so much @tomhoule. I'll slowly go through what I can do.

This is very helpful.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
